R/library.R

```diff
@@ -25,123 +25,86 @@ exists_in_lockbox <- function(locked_package) {
-    contrib.url(repos = "http://cran.us.r-project.org", type = "source"))
-  available <- data.frame(unique(available[, c("Package", "Version")]))
-  pkg <- available[available$Package == name, ]
-
-  # Simply install.packages if version happens to be the latest available on CRAN.
-  # You can specify the fastest CRAN mirror by setting the `lockbox.CRAN_mirror` option
-  # or Rstudio mirror will be used by default.
-  repos <- getOption('lockbox.CRAN_mirror') %||% c(CRAN = "http://cran.rstudio.com")
-  remote_version <- package_version(as.character(pkg$Version))
-  if (dim(pkg)[1] == 1 && remote_version == version) {
-    return(utils::install.packages(
-      name, repos = repos, INSTALL_opts = "--vanilla",
-      quiet = notTRUE(getOption('lockbox.verbose'))))
-  }
+install_package.CRAN <- function(locked_package, libP, quiet) {
+  filepath <- locked_package$download_path
+  utils::install.packages(filepath, lib = libP, repos = NULL, type = "source",
+    INSTALL_opts = "--vanilla", quiet = quiet)
+  unlink(filepath)
+}
 
-  # If we did not find the package on CRAN - try CRAN archive.
-  from <- paste0(repo, "/src/contrib/Archive/", name, "/", name, "_", version, ".tar.gz")
-  pkg.tarball <- tempfile(fileext = ".tar.gz")
-  download.file(url = from, destfile = pkg.tarball)
+install_package.github <- function(locked_package, libP, quiet) {
+  stopifnot(is.element("repo", names(locked_package)))
 
-  # We need to switch directories to ensure no infinite loop happens when
-  # the .Rprofile calls lockbox::lockbox.
-  old_dir <- getwd()
-  on.exit(setwd(old_dir))
-  tmpdir <- file.path(tempdir(), "foo")
-  dir.create(tmpdir, FALSE, TRUE)
-  setwd(tmpdir)
+  filepath <- locked_package$download_path
+  parent_dir <- dirname(filepath)
 
-  utils::install.packages(pkg.tarball, repos = NULL, type = "source",
-                          INSTALL_opts = "--vanilla",
-                          quiet = notTRUE(getOption("lockbox.verbose")))
-  unlink(pkg.tarball)
-}
+  ## Be very careful here to get the extracted package directory
+  ## while maintaining cross-platform support
+  extracted_dir <- gsub(file.path(parent_dir, "")
+    , "", unzip(filepath, exdir = parent_dir)[1])
+  extracted_dir <- file.path(parent_dir
+    , gsub(paste0(.Platform$file.sep,".*"), "", extracted_dir))
 
-install_package.CRAN <- function(locked_package) {
-  # TODO: (RK) Fetch correct version? Support install from source?
-  locked_package$repo <- locked_package$repo %||% "http://cran.r-project.org"
-  install_locked_package(locked_package,
-    install_old_CRAN_package(locked_package$name, locked_package$version))
-}
+  if (!is.null(locked_package$subdir)) {
+    extracted_dir <- file.path(extracted_dir, locked_package$subdir)
+  }
 
-#' @importFrom devtools install_github
-install_package.github <- function(locked_package) {
-  stopifnot(is.element("repo", names(locked_package)))
+  ## It's absolutely necessary that we make these executable
```

Comment:
Apparently R doesn't care.  That's either a good thing or another example of R being flexible to the point of madness.